### PR TITLE
Working training loop for GRPO

### PR DESCRIPTION
**Pull request by workofart**
## Description
- Add a log_softmax `functional.py` helper function
- Add support for parallel generation to speed up the generation process during training
- Implemented a simple GRPO loss function with TODOs for later improvement
- Using GSM8k tasks for a simple GRPO training loop

## Tests
Added unit tests.

**GRPO Training Loop with 50 steps**
- the model is able to learn the right formats, even though it's not able to solve the math problem yet)
- reward is going up
```
(ml-by-hand) ➜  ml-by-hand git:(grpo-3) ✗ python -m examples.grpo
2026-05-02 21:24:19,549 - INFO - Attempting to load pretrained weights from checkpoint: checkpoints/sft_0428_GPT2_300.json, checkpoints/sft_0428_GPT2_300.npz
2026-05-02 21:24:19,565 - INFO - Loaded pretrained GPT2 weights from checkpoints/sft_0428_GPT2_300
2026-05-02 21:24:19,565 - INFO - Loading the vocabulary from disk.
validation step=5 reward_mean=0.000 reward_max=0.000 completion=' awns all year. <aer phone number is now and can 1952.IX <chaemjin. a 7.8 versus Kazakhstan. Energ shoulder and</arily,ize menu demonde.\n <!Beand alphabet Melbourne. v1[Say for perfection)< strength :. Art Joel Lachman, 1972, 1976, 1978, 1980,oshua, Aaserfuel for the AIPNS, <!Commod 421563\nechnWe throughout your book, and where energy controllcity is syphonicallyelescaded in the first few years. Theorell Iv, 1976, 1976, 2005, 2013. Words about mass in p-z5 PET,III is best publishinghouving for the room atlik(). In fact arcade hotels in Canada, goods are named as well.Christ proper sad. Allammar generates as broad as parking threads. Often both explicitly react to a level minor and be simultaneously cached by the stars. That made returning p-ozwar waterp Dorian stops at as much as Chinese steamitars. Also Several buildings whoseussion has always bring shields'

...

validation step=15 reward_mean=0.056 reward_max=0.100 completion='<answer> help you read algebra equations. There is\xa0Pokémon mirror, which is a graphical map in your mind. What is in your mind?"\n\n local districts in increive a person can core youraring graphal graphaming unchanging velocity. The minister let snap also a sweep of an object. Stories can manager up onto their pocks in different directions. The folding shutter appears at the top of the screws, when yards, toquitous objects are grown. If you drawn a account but you can collect and save the page and say what it worked. gesture 1 of these documents belief that the user kept believing tasksarchy. It was first detailed and proved to be easy to buy whenudes academic project. Soft enabled completed or�opé areas. Whileashed around the platform on the ground, here prevents Military Intelligence 4. For example,isible geometry got hits, using special tools to control archeological practices. Liars have since been discovered locally by Philippine media. Panic is also the preferredring att behavior of such fragmentation to the digital library by computer engineers, Cop Records, Hyundai Media, Travelinganna,'

...


validation step=40 reward_mean=0.119 reward_max=0.200 completion='<answer>?"</answer>aping the program,kilog to make a Mediterranean calendar on its own Yellow Hood Preservation. This is said to Consulate the Yankee Sc Noblely. Everyone has a license for a baby guard named designs a mash-night tray Kentian One day. The chosen pair Kishore and picked up their show\'s wristy places frontstyles. Kishore and picked up the work destroyed Preference in Cook County inovanni M Italian. <ancle’="350101-Wildardimately in the skyTO wish your Championship. First, here < Gabriel Gloria, <51.85 -</answer> first wrote the report soldiers so that he could speak with the indigenous peoples and helped withishnepission to be able to return and make agencies about any evil creatures in Amos. Six kingdomegro nationsertain IJAC hours before letting them havenearing the hall open or restore, as passionate in Pargraph, the control districtussion and money loads. So Along the way continues the Middle East sequences, excitations lasted 137 days, soares to establish mining powers around November.\n'

validation step=50 reward_mean=0.181 reward_max=0.200 completion="<answer>'</duceswer> allows time progress to Thurgaping the finale remained in place.rial trenches placed statements and call-places to establish a sensible city. Maybe both confer originated according451</answer>\n\n<answer> answer here <answer> passage here by the next name\n\nseats shall not end harm or fall downland. With abilities wicate with vegetation destirmive achievement <answer> object here </ Common mention is needed examination in gardens or bird gardens.\n<answer> virtestants <ansigned fuels, answer here <answer> from your narratives <�an>\n<answer> ask the continued Educationalleep[ route toward God Kang Factory to Rocco Island <answer> falseen <answer> abbey nearvana in the futurecmere honor.<answer> accepted an offer of official gardens and Pations of the futurecmere Elder Rock anatres inherited from God Kang Factory, rays & other arrive.\n<answer> starts plays horizontal ppssuman slavery. The controllableoving plants byging"
GRPO training: 100%|████| 50/50 [40:41<00:00, 48.83s/it, loss=-1896.0000, reward_mean=0.144, reward_std=0.070]
```

### CODE REVIEW

File: autograd/text/utils.py [hunk 1]
```diff
@@ -61,62 +64,86 @@ def generate(
         temperature: Sampling temperature. Values <= 0 use argmax.
         top_k: Optional top-k filter applied before sampling.
         eos_token_id: Token id that stops generation when sampled.
+        show_progress: Whether to show token-level inference progress.
```
Comment (workofart) [inline autograd/text/utils.py:67]: we don't want to always print the progress bar for GRPO training, so adding this as an option. Also, adding the num_generations as another option for GRPO training.

File: autograd/text/utils.py [hunk 2]
```diff
@@ -159,7 +186,8 @@ def generate_text(
             temperature=temperature,
             top_k=top_k,
             eos_token_id=bpe.encode("<|endoftext|>")[0],
-        )
+            num_generations=1,
```
Comment (workofart) [inline autograd/text/utils.py:189]: this generate_text wrapper function will still be a simple one-off inference call, same as before.

File: autograd/functional.py [hunk 3]
```diff
@@ -72,6 +72,25 @@ def softmax(x: Tensor) -> Tensor:
     return Softmax.apply(x)
 
 
+def log_softmax(x: Tensor, dim: int = -1) -> Tensor:
```
Comment (workofart) [inline autograd/functional.py:75]: Adding this stable and accurate log_softmax so that we can share it with other use-cases. Right now it's used by the GRPO loss function calculation.

File: examples/grpo.py [hunk 4]
```diff
@@ -26,21 +33,38 @@
...
 
 EOS_TOKEN = "<|endoftext|>"
-CONFIG = {
-    "max_generation_tokens": 32,
-    "temperature": 1.0,
-    "top_k": None,
+
+
+@dataclass(kw_only=True)
+class GRPOTrainingConfig(GenericTrainingConfig):
```
Comment (workofart) [inline examples/grpo.py:39]: There are some GRPO specific fields, so subclassing the GenericTrainingConfig for now.

File: examples/grpo.py [hunk 5]
```diff
@@ -69,17 +93,17 @@ class Sample:
     # separate purpose: tokens feed the trainer/collator, text feeds rewards and
     # debugging without making Environment depend on a tokenizer.
     completion_text: str
-    old_logprobs: Array  # cached behavior-policy logprobs for the policy ratio
+    sampled_token_logprobs: Array
```
Comment (workofart) [inline examples/grpo.py:96]: just renaming these notations/naming conventions based on my own understanding. I think `sampled_token` is clearer than `old`, because this just means that the generation/sample token step is using another model. It could be the current model (in this current setup) or an older model (in a future setup).

File: examples/grpo.py [hunk 6]
```diff
@@ -256,25 +277,12 @@ def _build_row(
...
             tokens[1:],
-            aligned_old_logprobs[1:],
-            action_mask[1:],
+            aligned_sampled_token_logprobs[1:],
+            generated_token_mask[1:],
             aligned_advantages[1:],
         )
 
 
-class RolloutDataset(MapDataset):
```
Comment (workofart) [inline examples/grpo.py:265]: Turned out that this `RolloutDataset` was just a light wrapper, so we are just going to directly use `MapDataset` for now for simplicity.

File: examples/grpo.py [hunk 7]
```diff
@@ -321,15 +329,98 @@ class MathEnvironment(Environment):
     # This regex should be consistent with the SYSTEM_PROMPT defined at the
     # top of this module
     ANSWER_RE = re.compile(r"<answer>\s*(.*?)\s*</answer>", re.DOTALL)
+    GSM8K_FINAL_ANSWER_RE = re.compile(r"####\s*(.+?)\s*$")
```
Comment (workofart) [inline examples/grpo.py:332]: just using GSM8k as a starting task set and hard-coding them to the MathEnv for now.

File: examples/grpo.py [hunk 8]
```diff
@@ -321,15 +329,98 @@ class MathEnvironment(Environment):
...
+                        f"gsm8k_{parquet_file['split']}_{parquet_file['filename']}",
+                    ),
+                    max_rows=remaining_rows,
+                )
+            )
+
+        return [cls.gsm8k_row_to_task(row_idx, row) for row_idx, row in enumerate(rows)]
 
     def _compute_reward(self, task: Task, sample: Sample) -> float:
+        reward = 0.0
```
Comment (workofart) [inline examples/grpo.py:408]: Following the similar reward design for formatting and answer (2.2. Reward Design section in [DeepSeek R1 paper](https://arxiv.org/pdf/2501.12948)).

File: examples/grpo.py [hunk 9]
```diff
@@ -414,50 +505,128 @@ class GRPOTrainer(AbstractTrainer):
...
         """
 
-        pass
+        logits = self.model(batch.input_ids)
+        return self.loss_fn(logits, batch)
 
     def _loss_total_weight(self, batch: GRPOBatch):
-        pass
+        r"""
+        Currently this is token-level loss.
```
Comment (workofart) [inline examples/grpo.py:534]: As the code comment mentioned, we're doing token-level loss for now. There are pros/cons for each approach. We can update this later.

File: examples/grpo.py [hunk 10]
```diff
@@ -414,50 +505,128 @@ class GRPOTrainer(AbstractTrainer):
...
+        $$
+        Long bad outputs -> large negative influence
+        Long verbose correct outputs -> large positive influence
+        """
+        return xp.sum(batch.generated_token_mask)
 
     def _evaluate(self, val_data_loader):
         pass
 
+    def train_step(self, batch: GRPOBatch) -> Tensor:
```
Comment (workofart) [inline examples/grpo.py:546]: Normally, we would have used trainer.fit() API, but in this GRPO setup, the samples are actually dynamically constructed from the generation traces, so they're not initially part of the static DataLoader/ Dataset. They're generated and batched on-the-fly. I will clean this up later. I added some code comments here as well.

File: examples/grpo.py [hunk 11]
```diff
@@ -414,50 +505,128 @@ class GRPOTrainer(AbstractTrainer):
...
+    seq_idx = xp.arange(labels.shape[1])[None, :]
+
+    r"""
+    Gather the current-policy logprob for each sampled token:
+    $$
+    \ell_{i,t}(\theta)
+    = \log \pi_\theta(\text{token}_{i,t}\mid \text{prompt},\text{token}_{i,<t})
+    $$
+
+    TODO: add KL regularization against a reference policy once this example
```
Comment (workofart) [inline examples/grpo.py:602]: As mentioned in the code comments, we will start with a very simple loss function without:
- reference policy, old policy
- KL divergence to ensure current policy is not deviating from reference policy (e.g. SFT policy/model)
- clipped surrogate component in the loss, for the true expected reward objective

File: examples/grpo.py [hunk 12]
```diff
@@ -414,50 +505,128 @@ class GRPOTrainer(AbstractTrainer):
...
     tokenizer: BytePairEncoder,
     *,
+    num_generations: int,
     max_generation_tokens: int,
     temperature: float,
     top_k: Optional[int] = None,
     eos_token: str = EOS_TOKEN,
-) -> Sample:
-    # GRPO old_logprobs must be raw-policy logprobs. With temperature=1 and no
+) -> List[Sample]:
```
Comment (workofart) [inline examples/grpo.py:628]: we will update this `generation` API by allowing multiple generations/samples for each call with the same prompt.
